modin/engines/base/frame/partition.py

```diff
@@ -136,17 +140,55 @@ def mask(self, row_indices, col_indices):
 
         Parameters
         ----------
-        row_indices : list-like
+        row_indices : list-like, slice or label
             The indices for the rows to extract.
-        col_indices : list-like
+        col_indices : list-like, slice or label
             The indices for the columns to extract.
 
         Returns
         -------
         PandasFramePartition
             New `PandasFramePartition` object.
         """
-        pass
+
+        def is_full_axis_mask(index, axis_length):
+            """Check whether `index` mask grabs `axis_length` amount of elements."""
+            if isinstance(index, slice):
+                return index == slice(None) or (
+                    isinstance(axis_length, int)
+                    and compute_sliced_len(index, axis_length) == axis_length
+                )
+            return (
+                hasattr(index, "__len__")
+                and isinstance(axis_length, int)
+                and len(index) == axis_length
+            )
+
+        row_indices = [row_indices] if is_scalar(row_indices) else row_indices
+        col_indices = [col_indices] if is_scalar(col_indices) else col_indices
+
+        if is_full_axis_mask(row_indices, self._length_cache) and is_full_axis_mask(
+            col_indices, self._width_cache
+        ):
+            return copy(self)
+
+        new_obj = self.add_to_apply_calls(lambda df: df.iloc[row_indices, col_indices])
+
+        def try_recompute_cache(indices, previous_cache):
+            """Compute new axis-length cache for the masked frame based on its previous cache."""
+            return (
+                (
+                    compute_sliced_len(indices, previous_cache)
+                    if isinstance(previous_cache, int)
+                    else None
+                )
+                if isinstance(indices, slice)
+                else len(indices)
+            )
```

Comment:
maybe unwrap to a full `if..else`?

```suggestion
            if not isinstance(indices, slice):
                return len(indices)
            if not isinstance(previous_cache, int):
                return None
            return compute_sliced_len(indices, previous_cache)
```

looks shorter and is (hopefully) more readable

Reply:
applied